groups/service.go

```diff
@@ -0,0 +1,415 @@
+	}
+
+	return &groupService{client: client}, nil
+}
+
+type adminService struct {
+	client AdminServiceClient
+	group  GoogleGroup
+}
+
+func (as *adminService) SetGroup(group GoogleGroup) {
+	as.group = group
+}
+
+func (as *adminService) AddOrUpdateGroupMembers(role string, members []string) error {
+	if *verbose {
+		log.Printf("addOrUpdateMembersToGroup %s %q %v", role, as.group.EmailId, members)
+	}
+
+	l, err := as.client.ListMembers(as.group.EmailId)
+	if err != nil {
+		if apierr, ok := err.(*googleapi.Error); ok && apierr.Code == http.StatusNotFound {
+			log.Printf("skipping adding members to group %q as it has not yet been created\n", as.group.EmailId)
+			return nil
+		}
+		return fmt.Errorf("unable to retrieve members in group %q: %v", as.group.EmailId, err)
+	}
+
+	for _, memberEmailId := range members {
+		var member *admin.Member
+		for _, m := range l.Members {
+			if m.Email == memberEmailId {
+				member = m
+				break
+			}
+		}
+
+		if member != nil {
+			// update if necessary
+			if member.Role != role {
+				member.Role = role
+				if config.ConfirmChanges {
+					_, err := as.client.UpdateMember(as.group.EmailId, member.Email, member)
+					if err != nil {
+						return fmt.Errorf("unable to update %s in %q as %s : %v", memberEmailId, as.group.EmailId, role, err)
+					}
+					log.Printf("Updated %s to %q as a %s\n", memberEmailId, as.group.EmailId, role)
+				} else {
+					log.Printf("dry-run: would update %s to %q as %s\n", memberEmailId, as.group.EmailId, role)
+				}
+			}
+			continue
+		}
+		member = &admin.Member{
+			Email: memberEmailId,
+			Role:  role,
+		}
+
+		// We did not find the person in the google group, so we add them
+		if config.ConfirmChanges {
+			_, err := as.client.InsertMember(as.group.EmailId, member)
+			if err != nil {
+				return fmt.Errorf("unable to add %s to %q as %s : %v", memberEmailId, as.group.EmailId, role, err)
+			}
+			log.Printf("Added %s to %q as a %s\n", memberEmailId, as.group.EmailId, role)
+		} else {
+			log.Printf("dry-run: would add %s to %q as %s\n", memberEmailId, as.group.EmailId, role)
+		}
+	}
+
+	return nil
+}
+
+func (as *adminService) CreateOrUpdateGroupIfNescessary() error {
+	if *verbose {
+		log.Printf("createOrUpdateGroupIfNecessary %q", as.group.EmailId)
+	}
+
+	group, err := as.client.GetGroup(as.group.EmailId)
+	if err != nil {
+		if apierr, ok := err.(*googleapi.Error); ok && apierr.Code == http.StatusNotFound {
+			if !config.ConfirmChanges {
+				log.Printf("dry-run: would create group %q\n", as.group.EmailId)
+			} else {
+				log.Printf("Trying to create group: %q\n", as.group.EmailId)
+				g := admin.Group{
+					Email: as.group.EmailId,
+				}
+				if as.group.Name != "" {
+					g.Name = as.group.Name
+				}
+				if as.group.Description != "" {
+					g.Description = as.group.Description
+				}
+				g4, err := as.client.InsertGroup(&g)
+				if err != nil {
+					return fmt.Errorf("unable to add new group %q: %v", as.group.EmailId, err)
```

Comment:
Out of scope for this PR, but for followup: it would be ideal if we used `%w` instead of `%v` for errs